lib/modules.nix

```diff
@@ -342,28 +347,52 @@ rec {
       };
 
       result = withWarnings {
+        _type = "configuration";
         options = checked options;
         config = checked (removeAttrs config [ "_module" ]);
         _module = checked (config._module);
         inherit extendModules type;
       };
     in result;
 
-  # collectModules :: (modulesPath: String) -> (modules: [ Module ]) -> (args: Attrs) -> [ Module ]
+  # collectModules :: (class: String) -> (modulesPath: String) -> (modules: [ Module ]) -> (args: Attrs) -> [ Module ]
   #
   # Collects all modules recursively through `import` statements, filtering out
   # all modules in disabledModules.
-  collectModules = let
+  collectModules = class: let
 
       # Like unifyModuleSyntax, but also imports paths and calls functions if necessary
       loadModule = args: fallbackFile: fallbackKey: m:
-        if isFunction m || isAttrs m then
+        if isFunction m then
           unifyModuleSyntax fallbackFile fallbackKey (applyModuleArgsIfFunction fallbackKey m args)
+        else if isAttrs m then
+          if m._type or "module" == "module" then
+            unifyModuleSyntax fallbackFile fallbackKey (applyModuleArgsIfFunction fallbackKey m args)
+          else if m._type == "if" || m._type == "override" then
+            loadModule args fallbackFile fallbackKey { config = m; }
```

Comment:
This is new functionality, I'm not sure I like this idea. For readers of this comment, this is the associated test case for a module that now works:
```nix
{
  imports = [ (lib.mkIf true { enable = true; }) ];
}
```

This adds another type of module that's valid, in addition to files, attribute sets with specific keys, attribute sets with arbitrary keys (implicit `config`), functions returning such attribute sets.

It's also a bit weird in that `mkIf` and `mkForce` (and co.) are only available under `lib`, which wouldn't even be in scope in e.g. a standalone file containing such a module. So this module type is strictly only useful when you already have a `lib` in scope (this also irks me a bit because `lib` contains part of the module system API, and it potentially leads [to problems](https://github.com/NixOS/nixpkgs/pull/98952) if you mix them).

Reply:
This isn't new functionality; the test I added already worked before this change. Double checked :)

> irks

We should consider `mkIf` and friends to be a stable interface by now. They are little more than constructors, and I see no reason to change them in the future, unless we want to replace them by something that isn't the module system anymore. For what it's worth; this is basically off-topic considering there's no change in behavior here.

Reply:
> This isn't new functionality; the test I added already worked before this change. Double checked :)

Ahh sorry yes! Yeah nevermind :)